gubernator/main.py

```diff
@@ -0,0 +1,198 @@
+            data = memcache.get(key, namespace=namespace)
+            if data is not None:
+                return data
+            else:
+                data = func(arg)
+                if data:
+                    memcache.add(key, data, exp, namespace=namespace)
+                else:
+                    memcache.add(key, data, neg_exp, namespace=namespace)
+                return data
+        return wrapped
+    return wrapper
+
+
+@memcache_memoize('gs://')
+def gcs_read(path):
+    try:
+        with gcs.open(path) as f:
+            return f.read()
+    except gcs.errors.Error:
+        return None
+
+
+@memcache_memoize('build-details://', exp=60 * 60 * 4)
+def build_details(build_dir):
+    started = gcs_read(build_dir + '/started.json')
+    finished = gcs_read(build_dir + '/finished.json')
+    if not (started and finished):
+        return
+    started = json.loads(started)
+    finished = json.loads(finished)
+    failures = []
+    for n in xrange(1, 99):
+        junit = gcs_read(
+            '%s/artifacts/junit_%02d.xml' % (build_dir, n))
+        if junit is None:
+            break
+        failures.extend(parse_junit(decompress(junit)))
+    return started, finished, failures
+
+
+def decompress(data):
+    if data.startswith('\x1f\x8b'):  # gzip magic
+        return zlib.decompress(data, 15 | 16)
+    return data
+
+
+def parse_junit(xml):
+    for child in ET.fromstring(xml):
+        name = child.attrib['name']
+        time = float(child.attrib['time'])
+        failed = False
+        skipped = False
+        text = None
+        for param in child:
+            if param.tag == 'skipped':
+                skipped = True
+                text = param.text
+            elif param.tag == 'failure':
+                failed = True
+                text = param.text
+        if failed:
+            yield name, time, text
+
+
+class BuildHandler(webapp2.RequestHandler):
+    def get(self, bucket, prefix, job, build):
+        if bucket not in BUCKET_WHITELIST:
+            self.error(404)
+            return
+        build_dir = '/%s/%s%s/%s' % (bucket, prefix, job, build)
+        details = build_details(build_dir)
+        if not details:
+            self.error(404)
+            return
+        started, finished, failures = details
+        commit = started['version'].split('+')[1]
+        template = JINJA_ENVIRONMENT.get_template('build.html')
+        self.response.write(template.render(dict(
+            build_dir=build_dir, job=job, build=build, commit=commit,
+            started=started, finished=finished, failures=failures)))
+
+app = webapp2.WSGIApplication([
+    (r'/build/([-\w]+)/([-\w/]*/)?([-\w]+)/(\d+)/?', BuildHandler),
+    # webapp2.Route('/upload', gcs_upload.Upload, 'upload'),
+], debug=True)
+
+if os.environ.get('SERVER_SOFTWARE','').startswith('Development'):
+    # inject some test data so there's a page with some content
+    import tarfile
+    tf = tarfile.open('kube_results.tar.gz')
```

Comment:
Can you ask around and see if there are policies about storing binary data in our repo? I'm not sure the answer to this question (but neat way to speed up local dev btw!)